### DISCUSSION TITLE
Proposal: Use `Variant` type instead of `Array` type and `Object` Type

### DISCUSSION THREAD

**Discussion by andylokandy** [databendlabs/databend#8400] [General]
`Array` type and `Object` type are both `Variant` type internally but with the constraint of being `jsonb::Array` and `jsonb::Object`. Since the constraint doesn't affect any operation on variants, I propose to remove these two types.

**Comment by andylokandy** [comment]
cc @sundy-li @b41sh @BohuTANG

**Reply by sundy-li** [reply]
`Variant` is a dynamic type but `Array<T>` is a fixed type if `T` is already a fixed type. I think it would be better if we want to keep the exactly inner type when using Array Type.


Postgresql also keeps these three datatypes:  `Array Type`,  `Json Type`, `Composite Types` (Object type)

https://www.postgresql.org/docs/current/datatype.html

**Reply by andylokandy** [reply]
sorry, maybe my statement was confusing. I'm talking about the VariantObject and VariantArray.

**Reply by sundy-li** [reply]
> sorry, maybe my statement was confusing. I'm talking about the VariantObject and VariantArray.

Yes, that's ok!  We don't need `VariantObject`, `VariantArray` any more.
